### PR TITLE
Clinical behavior, historical notes, assignment, chemistry, fostering and staff ETLs

### PR DESCRIPTION
**Pull request by labkey-martyp**
#### Rationale
Clinical behavior, historical notes, assignment, chemistry, fostering and staff ETLs.  Performed by column matches on first/last name instead of email. Fix protocol and assignment ETLs.

#### Changes
* Fix protocol ETL query
* Issue 44509: Add incremental to assignment ETL and fix dupes
* Add clinical behavior to obs ETL
* Create notes, chemistry, fostering and staff ETLs
* Issue 44508: Add new column transform class for performedBy columns. Finding user by first/last name instead of email.
* Animal history case sensitive filter fix for participant view

### CODE REVIEW

File: nirc_ehr/resources/queries/dbo/q_chemistryResults.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+       anmCmt.TEXT                                                               AS remark,
+       CAST(COALESCE(adt.modified, anmEvt.CREATED_DATETIME) AS TIMESTAMP)        AS modified
+FROM ANIMAL_EVENT anmEvt
+         LEFT JOIN ANIMAL_EVENT_COMMENT anmCmt ON anmEvt.ANIMAL_EVENT_ID = anmCmt.ANIMAL_EVENT_ID
+         LEFT JOIN EVENT_EVENT_GROUP evtEvtGrp ON evtEvtGrp.EVENT_ID = anmEvt.EVENT_ID
+         LEFT JOIN q_modified_event adt ON anmEvt.ANIMAL_EVENT_ID = adt.event_id
+
+WHERE evtEvtGrp.EVENT_GROUP_ID = 64
+  AND anmEvt.CREATED_DATETIME < now() -- there are rows in ANIMAL_EVENT table with future dates
+--     64 Hormones
```
Comment (labkey-jeckels) [inline nirc_ehr/resources/queries/dbo/q_chemistryResults.sql]: Can this comment be deleted?

File: nirc_ehr/resources/queries/dbo/q_foster.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+       anmCmt.TEXT                                                               AS remark,
+       CAST(COALESCE(adt.modified, anmEvt.CREATED_DATETIME) AS TIMESTAMP)        AS modified
+FROM ANIMAL_EVENT anmEvt
+         LEFT JOIN ANIMAL_EVENT_COMMENT anmCmt ON anmEvt.ANIMAL_EVENT_ID = anmCmt.ANIMAL_EVENT_ID
+         LEFT JOIN EVENT_EVENT_GROUP evtEvtGrp ON evtEvtGrp.EVENT_ID = anmEvt.EVENT_ID
+         LEFT JOIN q_modified_event adt ON anmEvt.ANIMAL_EVENT_ID = adt.event_id
+
+WHERE evtEvtGrp.EVENT_GROUP_ID = 38
+  AND anmEvt.CREATED_DATETIME < now() -- there are rows in ANIMAL_EVENT table with future dates
+--     64 Hormones
```
Comment (labkey-jeckels) [inline nirc_ehr/resources/queries/dbo/q_foster.sql]: Or should this be explaining what EVENT_GROUP_ID = 38 is filtering on instead?

File: nirc_ehr/src/org/labkey/nirc_ehr/columnTransform/NIRCUserCreateTransform.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class NIRCUserCreateTransform extends ColumnTransform
+{
+    private transient Map<String, Integer> _sourceUserNameMap = new HashMap<>();
+
+    /**
+     * Search for the user by the email prefix (ie. jdoe@labkey.com matches jdoe).
```
Comment (labkey-jeckels) [inline nirc_ehr/src/org/labkey/nirc_ehr/columnTransform/NIRCUserCreateTransform.java]: Comment doesn't match the implementation - it's using name, not email username

File: nirc_ehr/src/org/labkey/nirc_ehr/columnTransform/NIRCUserCreateTransform.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        new SqlSelector(us.getDbSchema(), sql).forEachMap(row -> {
+            String displayName;
+            if (row.get("displayName") != null)
+            {
+                displayName = (String)row.get("displayName");
+            }
+            else
+            {
+                String EMAIl_PATTERN = "[^a-zA-Z0-9!#$%&@'*+-/=?^_`{|}~.]+";
+                displayName = (nameParts[0] + nameParts[1]).replaceAll(EMAIl_PATTERN, "");
```
Comment (labkey-jeckels) [inline nirc_ehr/src/org/labkey/nirc_ehr/columnTransform/NIRCUserCreateTransform.java]:
```suggestion
                String EMAIL_PATTERN = "[^a-zA-Z0-9!#$%&@'*+-/=?^_`{|}~.]+";
                displayName = (nameParts[0] + nameParts[1]).replaceAll(EMAIL_PATTERN, "");
```

File: nirc_ehr/resources/schemas/dbscripts/postgresql/nirc_ehr-21.023-22.000.sql [hunk 5]
```diff
@@ -0,0 +1,78 @@
+CREATE TABLE nirc_ehr.Staff
```
Comment (labkey-jeckels) [inline nirc_ehr/resources/schemas/dbscripts/postgresql/nirc_ehr-21.023-22.000.sql]: I don't see an edit to NIRC_EHRModule.java to bump the schema version. You probably need to go to 22.001 with the back port of the schema version bump from develop.

**Review by labkey-jeckels — APPROVED** [review]
Small edits but no need for re-review unless you want one.

File: nirc_ehr/resources/schemas/dbscripts/postgresql/nirc_ehr-21.023-22.000.sql (hunk 5 above)
Comment (labkey-adam) [reply nirc_ehr/resources/schemas/dbscripts/postgresql/nirc_ehr-21.023-22.000.sql]: Right, and the script probably needs to be named `nirc_ehr-22.000-22.001.sql` to run on all dev/test machines.

Comment (labkey-martyp) [reply nirc_ehr/resources/schemas/dbscripts/postgresql/nirc_ehr-21.023-22.000.sql]: Fixed this. Rushed out the PR before I tested the startup script.

File: nirc_ehr/src/org/labkey/nirc_ehr/columnTransform/NIRCUserCreateTransform.java (hunk 4 above)
Comment (labkey-martyp) [reply nirc_ehr/src/org/labkey/nirc_ehr/columnTransform/NIRCUserCreateTransform.java]: Fixed

File: nirc_ehr/src/org/labkey/nirc_ehr/columnTransform/NIRCUserCreateTransform.java (hunk 3 above)
Comment (labkey-martyp) [reply nirc_ehr/src/org/labkey/nirc_ehr/columnTransform/NIRCUserCreateTransform.java]: Good catch, fixed.

File: nirc_ehr/resources/queries/dbo/q_foster.sql (hunk 2 above)
Comment (labkey-martyp) [reply nirc_ehr/resources/queries/dbo/q_foster.sql]: Updated to correct group in comment

File: nirc_ehr/resources/queries/dbo/q_chemistryResults.sql (hunk 1 above)
Comment (labkey-martyp) [reply nirc_ehr/resources/queries/dbo/q_chemistryResults.sql]: I moved it up, closer to its context

**Pull request merged**